In [2]:
# Install packages
!pip install -r requirements.txt >/dev/null
!pip install eep153_tools >/dev/null
!pip install gnupg 
!pip install gspread_pandas >/dev/null

# Import libraries
import numpy as np
import pandas as pd
import warnings
import fooddatacentral as fdc
from  scipy.optimize import linprog as lp
import plotly.express as px
from eep153_tools.sheets import read_sheets

# API Key
apikey = "Y7y8evhwYYL589qVEZzdHBCyKGeBkNdDWQfnWyIV"



  Using cached gnupg-2.3.1-py3-none-any.whl


In [94]:
# Read diet minimum data
diet_min = pd.read_csv("Spreadsheets/diet_min.csv",index_col=0)
diet_min

Source   C 1-3   F 4-8   M 4-8  F 9-13  M 9-13  \
Nutrition                                                                       
Energy                            ---  1000.0  1200.0  1400.0  1600.0  1800.0   
Protein                           RDA    13.0    19.0    19.0    34.0    34.0   
Fiber, total dietary              ---    14.0    16.8    19.6    22.4    25.2   
Folate, DFE                       RDA   150.0   200.0   200.0   300.0   300.0   
Calcium, Ca                       RDA   700.0  1000.0  1000.0  1300.0  1300.0   
Carbohydrate, by difference       RDA   130.0   130.0   130.0   130.0   130.0   
Iron, Fe                          RDA     7.0    10.0    10.0     8.0     8.0   
Magnesium, Mg                     RDA    80.0   130.0   130.0   240.0   240.0   
Niacin                            RDA     6.0     8.0     8.0    12.0    12.0   
Phosphorus, P                     RDA   460.0   500.0   500.0  1250.0  1250.0   
Potassium, K                       AI  3000.0  3800.0  3800.0  4500.0  4500.0   
Riboflavin                        RDA     0.5     0.6     0.6     0.9     0.9   
Thiamin                           RDA     0.5     0.6     0.6     0.9     0.9   
Vitamin A, RAE                    RDA   300.0   400.0   400.0   600.0   600.0   
Vitamin B-12                      RDA     0.9     1.2     1.2     1.8     1.8   
Vitamin B-6                       RDA     0.5     0.6     0.6     1.0     1.0   
Vitamin C, total ascorbic acid    RDA    15.0    25.0    25.0    45.0    45.0   
Vitamin E (alpha-tocopherol)      RDA     6.0     7.0     7.0    11.0    11.0   
Vitamin K (phylloquinone)          AI    30.0    55.0    55.0    60.0    60.0   
Zinc, Zn                          RDA     3.0     5.0     5.0     8.0     8.0   

                                F 14-18  M 14-18  F 19-30  M 19-30  F 31-50  \
Nutrition                                                                     
Energy                           1800.0   2200.0   2000.0   2400.0   1800.0   
Protein                            46.0     52.0     46.0     56.0     46.0   
Fiber, total dietary               25.2     30.8     28.0     33.6     25.2   
Folate, DFE                       400.0    400.0    400.0    400.0    400.0   
Calcium, Ca                      1300.0   1300.0   1000.0   1000.0   1000.0   
Carbohydrate, by difference       130.0    130.0    130.0    130.0    130.0   
Iron, Fe                           15.0     11.0     18.0      8.0     18.0   
Magnesium, Mg                     360.0    410.0    310.0    400.0    320.0   
Niacin                             14.0     16.0     14.0     16.0     14.0   
Phosphorus, P                    1250.0   1250.0    700.0    700.0    700.0   
Potassium, K                     4700.0   4700.0   4700.0   4700.0   4700.0   
Riboflavin                          1.0      1.3      1.1      1.3      1.1   
Thiamin                             1.0      1.2      1.1      1.2      1.1   
Vitamin A, RAE                    700.0    900.0    700.0    900.0    700.0   
Vitamin B-12                        2.4      2.4      2.4      2.4      2.4   
Vitamin B-6                         1.2      1.3      1.3      1.3      1.3   
Vitamin C, total ascorbic acid     65.0     75.0     75.0     90.0     75.0   
Vitamin E (alpha-tocopherol)       15.0     15.0     15.0     15.0     15.0   
Vitamin K (phylloquinone)          75.0     75.0     90.0    120.0     90.0   
Zinc, Zn                            9.0     11.0      8.0     11.0      8.0   

                                M 31-50   F 51+   M 51+  
Nutrition                                                
Energy                           2200.0  1600.0  2000.0  
Protein                            56.0    46.0    56.0  
Fiber, total dietary               30.8    22.4    28.0  
Folate, DFE                       400.0   400.0   400.0  
Calcium, Ca                      1000.0  1200.0  1000.0  
Carbohydrate, by difference       130.0   130.0   130.0  
Iron, Fe                            8.0     8

In [95]:
# Read diet maximum data
diet_max = pd.read_csv("Spreadsheets/diet_max.csv",index_col=0)
diet_max

Source  C 1-3  F 4-8  M 4-8  F 9-13  M 9-13  F 14-18  M 14-18  \
Nutrition                                                                  
Sodium, Na     UL   1500   1900   1900    2200    2200     2300     2300   
Energy        NaN   2500   2500   2500    2800    3000     3100     3100   

            F 19-30  M 19-30  F 31-50  M 31-50  F 51+  M 51+  
Nutrition                                                     
Sodium, Na     2300     2300     2300     2300   2300   2300  
Energy         3100     3100     3100     3100   3100   3100

In [6]:
def dietary_ref_intake(age,sex,df):
    """Takes in age and sex, and returns the dietary reference intake for the chosen population"""

    if age <= 3:
        col = 'C 1-3'
    age_ranges = [(4,8),(9,13),(14,18),(19,30),(31,50),(50,100)]
    for age_range in age_ranges:
        if age >= age_range[0] and age <= age_range[1]:
            col = sex + ' ' + str(age_range[0]) + '-' + str(age_range[1])
    return pd.Series(df[col])  

In [7]:
# Example of minimum dietary requirements for a male aged 22
dietary_ref_intake(age=22,sex='M',df=diet_min)

Nutrition
Energy                            2400.0
Protein                             56.0
Fiber, total dietary                33.6
Folate, DFE                        400.0
Calcium, Ca                       1000.0
Carbohydrate, by difference        130.0
Iron, Fe                             8.0
Magnesium, Mg                      400.0
Niacin                              16.0
Phosphorus, P                      700.0
Potassium, K                      4700.0
Riboflavin                           1.3
Thiamin                              1.2
Vitamin A, RAE                     900.0
Vitamin B-12                         2.4
Vitamin B-6                          1.3
Vitamin C, total ascorbic acid      90.0
Vitamin E (alpha-tocopherol)        15.0
Vitamin K (phylloquinone)          120.0
Zinc, Zn                            11.0
Name: M 19-30, dtype: float64

In [8]:
# Example of maximum dietary requirements for a male aged 22
dietary_ref_intake(age=22,sex='M',df=diet_max)

Nutrition
Sodium, Na    2300
Energy        3100
Name: M 19-30, dtype: int64

## Identifying and Uploading FDC

In [96]:
df = pd.read_csv('Spreadsheets/grocery_names.csv')
df['Food'] = df['Item']
df

Item                    Food
0            Hass Avocado            Hass Avocado 
1                     Tofu                    Tofu
2            Lacinato Kale           Lacinato Kale
3                 Oat Milk                Oat Milk
4        Flat Leaf Parsley       Flat Leaf Parsley
5          Maradol Papayas         Maradol Papayas
6            Gemelli Pasta           Gemelli Pasta
7            Roma Tomatoes           Roma Tomatoes
8          Serrano Peppers         Serrano Peppers
9                 Cilantro                Cilantro
10          Corn Tortillas          Corn Tortillas
11                 Chorizo                 Chorizo
12          Pressed Juices          Pressed Juices
13                 Arugula                 Arugula
14                  Apples                  Apples
15                 Oranges                 Oranges
16                 Bananas                 Bananas
17                Potatoes                Potatoes
18     Tomatoes (Vine Red)     Tomatoes (Vine Red)
19                    Milk                    Milk
20                  Cheese                  Cheese
21  Chips (Lay's Original)  Chips (Lay's Original)
22                   Bread                   Bread
23                   Flour                   Flour
24                 Chicken                 Chicken
25            Ground Beef             Ground Beef 
26                    Eggs                    Eggs
27                  Onions                  Onions

In [92]:
# (IGNORE FOR PROJECT) idea for further automation/ease here: if you inputted a dataframe with your 
# chosen foods having already designated for the "foodCategory" column, that way you could have some 
# kind of if/else statement that filters to find the FIRST index row with that DESIRED (and more accurate) 
# food category as the value in that column of the FDC dataframe. This way you could weed out rando ones, 
# like when we get "Avocado Chunks" as the SECOND result in a search for Hass Avocadoes. 

import fooddatacentral as fdc
import warnings

apikey = 'Y7y8evhwYYL589qVEZzdHBCyKGeBkNdDWQfnWyIV'

def find_FDC_index(API_key, food_list, num_rows): 
    
    fdc_index_results = pd.DataFrame(columns=['Food', 'fdcId', 'description', 'foodCategory'])
    
    for food in food_list: 
        df_food = fdc.search(apikey, food)
        df_food_reduced = df_food[['fdcId', 'description', 'foodCategory']]
        df_food_reduced['Food'] = str(food)
        df_food_4_results = df_food_reduced.iloc[:num_rows]
        fdc_index_results = fdc_index_results.append(df_food_4_results, ignore_index = True)
        
    return fdc_index_results 

In [93]:
get_fdcIds = find_FDC_index(apikey, df['Item'], 1)
get_fdcIds

/tmp/ipykernel_82/3561763109.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_food_reduced['Food'] = str(food)


Food    fdcId  \
0            Hass Avocado   2609822   
1                     Tofu  2294522   
2            Lacinato Kale  2495872   
3                 Oat Milk  2257046   
4        Flat Leaf Parsley   170416   
5          Maradol Papayas   169926   
6            Gemelli Pasta  2618997   
7            Roma Tomatoes  1608947   
8          Serrano Peppers  2390015   
9                 Cilantro   399147   
10          Corn Tortillas  1850647   
11                 Chorizo  2341591   
12          Pressed Juices  2341550   
13                 Arugula  2630019   
14                  Apples  1545857   
15                 Oranges   578523   
16                 Bananas   173945   
17                Potatoes   576920   
18     Tomatoes (Vine Red)   170050   
19                    Milk  1909132   
20                  Cheese  2679762   
21  Chips (Lay's Original)  1635045   
22                   Bread  2431121   
23                   Flour  2671058   
24                 Chicken  2090362   
25            Ground Beef   1942314   
26                    Eggs  1663044   
27                  Onions  2345315   

                                          description  \
0                                 HASS AVOCADOS, HASS   
1                                                TOFU   
2   LACINATO TUSCAN KALE CUT SUPER GREENS, LACINAT...   
3          Oat milk, unsweetened, plain, refrigerated   
4                                      Parsley, fresh   
5                                        Papayas, raw   
6    GEMELLI & SPINACH PASTA SALAD, GEMELLI & SPINACH   
7                                       ROMA TOMATOES   
8                                     SERRANO PEPPERS   
9                                            CILANTRO   
10                                     CORN TORTILLAS   
11                                            Chorizo   
12                             Duck, pressed, Chinese   
13                                            ARUGULA   
14                                             APPLES   
15                                            ORANGES   
16              Bananas, dehydrated, or banana powder   
17                                           POTATOES   
18                        Tomatoes, red, ripe, cooked   
19                                               MILK   
20                                             CHEESE   
21       LAY'S, WAVY POTATO CHIPS, ORIGINAL, ORIGINAL   
22                                              BREAD   
23                                              FLOUR   
24                                            CHICKEN   
25                                        GROUND BEEF   
26                                               EGGS   
27                                        Onions, raw   

                           foodCategory  
0       Pre-Packaged Fruit & Vegetables  
1                           Other Meats  
2       Pre-Packaged Fruit & Vegetables  
3                             Beverages  
4     Vegetables and Vegetable Products  
5               Fruits and Fruit Juices  
6                           Deli Salads  
7       Pre-Packaged Fruit & Vegetables  
8   Pickles, Olives, Peppers & Relishes  
9                        Herbs & Spices  
10                 Mexican Dinner Mixes  
11                             Sausages  
12          Turkey, duck, other poultry  
13      Pre-Packaged Fruit & Vegetables  
14      Pre-Packaged Fruit & Vegetables  
15      Pre-Packaged Fruit & Vegetables  
16              Fruits and Fruit Juices  
17      Pre-Packaged Fruit & Vegetables  
18    Vegetables and Vegetable Products  
19                                 Milk  
20                               Cheese  
21             Chips, Pretzels & Snacks  
22                   Cookies & Biscuits  
23                   Flours & Corn Meal  
24            Poultry, Chicken & Turkey  
25                          Other Meats  
26               Eggs & Egg Substitutes  
27                               Onions

In [82]:
fdcId_array = get_fdcIds['fdcId']
print(fdcId_array)

0     2609822
1     2294522
2     2495872
3     2257046
4      170416
5      169926
6     2618997
7     1608947
8     2390015
9      399147
10    1850647
11    2341591
12    2341550
13    2630019
14    1545857
15     578523
16     173945
17     576920
18     170050
19    1909132
20    2679762
21    1635045
22    2431121
23    2671058
24    2090362
25    1942314
26    1663044
27    2345315
Name: fdcId, dtype: object


In [83]:
file_path = "fdc_ids.csv"

# Save the DataFrame to a CSV file
fdcId_array.to_csv(file_path, index=False)

### Notes from FDC ID DataFrame: 
- for Gemelli Pasta, pd.iloc[0] will get fdcId = 2618997, change to fdcId = 1124597
- "pressed Juices" are fucked up, need to do that search separately
- "Bread" is outputting some cookies and biscuits shit, replace with fdcId = 1913550

In [91]:
gemelli_fdcID = 1124597
bread_fdcID = 1913550
pressedjuice_fdcID = 2095092

In [90]:
# getting FDC Codes

import fooddatacentral as fdc
import warnings
df1 = fdc.search(apikey, 'Cold Pressed Juice')
df1

fdcId                      description   dataType       gtinUpc  \
0   2075597               COLD PRESSED JUICE    Branded  852043003810   
1    522238               COLD PRESSED JUICE    Branded  857134006749   
2   1919466               COLD PRESSED JUICE    Branded  653341767899   
3   2095092               COLD PRESSED JUICE    Branded  813377021864   
4   1889790               COLD PRESSED JUICE    Branded  854649005960   
5    523140               COLD PRESSED JUICE    Branded  857134006640   
6   1924431               COLD PRESSED JUICE    Branded  813377021932   
7   1917389               COLD PRESSED JUICE    Branded  813377021529   
8   1917763               COLD PRESSED JUICE    Branded  857134006350   
9   1918514               COLD PRESSED JUICE    Branded  813377021628   
10  2540936               COLD PRESSED JUICE    Branded  856650007117   
11  1930570               COLD PRESSED JUICE    Branded  813377022328   
12  1897029               COLD PRESSED JUICE    Branded  854649005977   
13  1917755               COLD PRESSED JUICE    Branded  857134006398   
14  1925436               COLD PRESSED JUICE    Branded  854649005649   
15   167702      Oil, flaxseed, cold pressed  SR Legacy           NaN   
16  2640714               COLD-PRESSED JUICE    Branded  854649005090   
17  2430838          100% COLD PRESSED JUICE    Branded  752830076280   
18  2427629          100% COLD PRESSED JUICE    Branded  752830078185   
19  2427627          100% COLD PRESSED JUICE    Branded  752830078086   
20  2426735          100% COLD PRESSED JUICE    Branded  752830077683   
21  1949152          100% COLD PRESSED JUICE    Branded  752830076181   
22  2430841          100% COLD PRESSED JUICE    Branded  752830076389   
23  1949872          100% COLD PRESSED JUICE    Branded  752830076082   
24  2434223          100% COLD PRESSED JUICE    Branded  752830597242   
25  1877450          100% COLD-PRESSED JUICE    Branded  811129020165   
26  1875943          100% COLD-PRESSED JUICE    Branded  811129020158   
27  1919003         COLD PRESSED APPLE JUICE    Branded  813377021543   
28  2337897         COLD PRESSED APPLE JUICE    Branded  681131093071   
29  2437319    COLD PRESSED GRAPEFRUIT JUICE    Branded  748753922292   
30  2075598      COLD PRESSED JUICE BEVERAGE    Branded  852043003834   
31   498418      COLD PRESSED JUICE BEVERAGE    Branded  852043003742   
32  2385794          COLD PRESSED JUICE SHOT    Branded  030034938987   
33  1927937   COLD PRESSED JUICE, GRAPEFRUIT    Branded  854649005670   
34  1931203     COLD PRESSED JUICE, LEMONADE    Branded  856650007247   
35  2640715       COLD PRESSED JUICE, ORANGE    Branded  854649005915   
36  2436737        COLD PRESSED JUICE, SWEET    Branded  856081007090   
37  1875120  COLD PRESSED JUICED, WATERMELON    Branded  852921005103   
38  1906348  COLD PRESSED JUICED, WATERMELON    Branded  852921005127   
39  2139509      COLD PRESSED LEMONADE JUICE    Branded   07789047452   
40  2039302       COLD PRESSED ORGANIC JUICE    Branded  840469110052   
41  2670122       COLD PRESSED SUPER A JUICE    Branded  856081007052   
42  1137698     COLD PRESSED TANGERINE JUICE    Branded      20990534   
43  1882202      COLD-PRESSED JUICE, RESTORE    Branded  791090770538   
44  2479405     COLD-PRESSED SPARKLING JUICE    Branded  818617024483   
45  1948784        GINGER COLD-PRESSED JUICE    Branded  091037462473   
46  2546078       ORGANIC COLD PRESSED JUICE    Branded  858997006013   
47  2628787       ORGANIC COLD-PRESSED JUICE    Branded  858997006044   
48  1932971       ORGANIC COLD-PRESSED JUICE    Branded  851481006407   
49  2338509       ORGANIC COLD-PRESSED JUICE    Branded  851481006384   

   publishedDate                           brandOwner               brandName  \
0     2021-10-28       The Hain Celestial Group, Inc.               BLUEPRINT   
1     2019-04-01                   WHOLE FOODS MARKET                     NaN   
2     2021-07-29           A.M.S. Manuf

# TO BE CONTINUED: Getting Nutrition Dataframes (now that FDC Ids are set) 